x-pack/filebeat/input/httpjson/input.go

```diff
@@ -149,8 +164,127 @@ func (in *httpjsonInput) createHTTPRequest(ctx context.Context, ri *requestInfo)
 	return req, nil
 }
 
+// processEventArray publishes an event for each object contained in the array. It returns the last object in the array and an error if any.
+func (in *HttpjsonInput) processEventArray(events []interface{}) (map[string]interface{}, error) {
+	var m map[string]interface{}
+	for _, t := range events {
+		switch v := t.(type) {
+		case map[string]interface{}:
+			m = v
+			d, err := json.Marshal(v)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to marshal %+v", v)
+			}
+			ok := in.outlet.OnEvent(makeEvent(string(d)))
+			if !ok {
+				return nil, errors.New("function OnEvent returned false")
+			}
+		default:
+			return nil, errors.Errorf("expected only JSON objects in the array but got a %T", v)
+		}
+	}
+	return m, nil
+}
+
+// getNextLinkFromHeader retrieves the next URL for pagination from the HTTP Header of the response
+func getNextLinkFromHeader(header http.Header, fieldName string, re *regexp.Regexp) (string, error) {
+	links, ok := header[fieldName]
+	if !ok {
+		return "", errors.Errorf("field %s does not exist in the HTTP Header", fieldName)
+	}
+	for _, link := range links {
+		matchArray := re.FindAllStringSubmatch(link, -1)
+		if len(matchArray) == 1 {
+			return matchArray[0][1], nil
+		}
+	}
+	return "", nil
+}
+
+// getRateLimit get the rate limit value if specified in the HTTP Header of the response
+func getRateLimit(header http.Header, rateLimit *RateLimit) (int64, error) {
+	if rateLimit != nil {
+		if rateLimit.Remaining != "" {
+			remaining := header.Get(rateLimit.Remaining)
+			if remaining == "" {
+				return 0, errors.Errorf("field %s does not exist in the HTTP Header, or is empty", rateLimit.Remaining)
+			}
+			m, err := strconv.ParseInt(remaining, 10, 64)
+			if err != nil {
+				return 0, errors.Wrapf(err, "failed to parse rate-limit remaining value")
+			}
+			if m == 0 {
+				reset := header.Get(rateLimit.Reset)
+				if reset == "" {
+					return 0, errors.Errorf("field %s does not exist in the HTTP Header, or is empty", rateLimit.Reset)
+				}
+				epoch, err := strconv.ParseInt(reset, 10, 64)
+				if err != nil {
+					return 0, errors.Wrapf(err, "failed to parse rate-limit reset value")
+				}
+				return epoch, nil
+			}
+		}
+	}
+	return 0, nil
+}
+
+// applyRateLimit applies appropriate rate limit if specified in the HTTP Header of the response
+func (in *HttpjsonInput) applyRateLimit(ctx context.Context, header http.Header, rateLimit *RateLimit) error {
+	epoch, err := getRateLimit(header, rateLimit)
+	if err != nil {
+		return err
+	}
+	if epoch == 0 {
+		return nil
+	}
+	t := time.Unix(epoch, 0)
+	in.log.Debugf("Rate Limit: Wait until %v for the rate limit to reset.", t)
+	ticker := time.NewTicker(time.Until(t))
+	defer ticker.Stop()
+	for {
```

Comment:
Why does this need the for loop? Looks like both cases will return so only one loop is possible.

Reply:
You are right. Copy pasted from the other one. :-)